core/storage/question/gae_models.py

```diff
@@ -152,24 +152,62 @@ class QuestionSkillLinkModel(base_models.BaseModel):
     skill_id = ndb.StringProperty(required=True, indexed=True)
 
     @classmethod
-    def create(
-            cls, question_id, skill_id):
+    def get_model_id(cls, question_id, skill_id):
+        """Returns the model id by combining the questions and skill id.
+
+        Args:
+            question_id: str. The ID of the question.
+            skill_id: str. The ID of the skill to which the question is linked.
+
+        Returns:
+            str. The calculated model id.
+        """
+        return question_id + ':' + skill_id
+
+    @classmethod
+    def create(cls, question_id, skill_id):
         """Creates a new QuestionSkillLinkModel entry.
 
         Args:
             question_id: str. The ID of the question.
             skill_id: str. The ID of the skill to which the question is linked.
 
+        Raises:
+            Exception. The given question is already linked to the given skill.
+
         Returns:
             QuestionSkillLinkModel. Instance of the new QuestionSkillLinkModel
                 entry.
         """
+        question_skill_link_id = cls.get_model_id(question_id, skill_id)
+        if cls.get(question_skill_link_id, strict=False) is not None:
+            raise Exception(
+                'The given question is already linked to given skill')
 
         question_skill_link_model_instance = cls(
-            question_id=question_id, skill_id=skill_id)
-
+            id=question_skill_link_id,
+            question_id=question_id,
+            skill_id=skill_id
+        )
         return question_skill_link_model_instance
 
+    @classmethod
+    def get_question_ids_linked_to_skill_ids(cls, skill_ids):
+        """Creates a new QuestionSkillLinkModel entry.
+
+        Args:
+            skill_ids: list(str). The ids of skills for which the linked
+                question ids are to be retrieved.
+
+        Returns:
+            list(str). The question ids linked to given skills.
+        """
+        question_skill_link_models = cls.query(cls.skill_id.IN(skill_ids))
```

Comment:
I think you are going to run into some issues here. Try and do some experiments locally with 1000 models and 100 skill IDs -- you're going to be doing basically 100 queries, one for each skill ID. (I think this is the case based on https://cloud.google.com/appengine/docs/standard/python/ndb/queries#neq_and_in.)

So ... you should probably limit the number of skill IDs that can be passed into here, based on what's actually feasible, and enforce this limit. Maybe you can set it at 1 or 2 for now. You'll need to do the above-mentioned tests on prod datastore to see what works and what doesn't (consider filing an issue for this) -- maybe talk to @ishucr7 for advice on setting up a test prod server.

If it doesn't scale then we need to be clever -- maybe precompute question IDs for particular skill combinations via cron jobs or something. Let's discuss this when we meet next.

Reply:
OK, please link to it here once you've done that.

Reply:
Done, created [issue](https://github.com/oppia/oppia/issues/5343)